File: ios/ExpoClickOutsideModule.swift
import ExpoModulesCore
import UIKit

public class ExpoClickOutsideModule: Module {
  private let tapHandler = GlobalTapHandler()

  public func definition() -> ModuleDefinition {
    Name("ExpoClickOutside")

    Events("onTap")

    OnStartObserving {
      DispatchQueue.main.async {
        guard let window = UIApplication.shared.connectedScenes
          .compactMap({ $0 as? UIWindowScene })
          .flatMap({ $0.windows })
          .first(where: { $0.isKeyWindow }) else {
            print("ExpoClickOutsideModule: No active key window found")
            return
          }

        self.tapHandler.attach(to: window)
          self.tapHandler.onTap = { point, isTextInput, viewClass in
          self.sendEvent("onTap", [
            "x": point.x,
            "y": point.y,
            "targetIsTextInput": isTextInput,
            "target": viewClass
          ])
        }
      }
    }

    OnStopObserving {
       self.tapHandler.detach()
    }
  }
}

File: android/src/main/java/expo/modules/clickoutside/ExpoClickOutsideModule.kt
package expo.modules.clickoutside

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoClickOutsideModule : Module() {
  private var tapHandler: GlobalTapHandler? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoClickOutside")

    Events("onTap")

    OnStartObserving {
      val activity = appContext.currentActivity ?: return@OnStartObserving

      tapHandler = GlobalTapHandler { x, y, isTextInput, viewClass ->
        sendEvent("onTap", mapOf(
          "x" to x,
          "y" to y,
          "targetIsTextInput" to isTextInput,
          "target" to viewClass
        ))
      }

      tapHandler?.attach(activity)
    }

    OnStopObserving {
      val activity = appContext.currentActivity ?: return@OnStopObserving
      tapHandler?.detach(activity)
      tapHandler = null
    }
  }
}
